### DISCUSSION TITLE
How to store clients outside of the heap?

### DISCUSSION THREAD

**Discussion by Schuyweiz** [tdlight-team/tdlight-java#88] [Q&A]
I was wondering if it is possible to preserve client somewhere else besides the heap? 

Maybe I can somehow restore a client by client id. Or maybe I can somehow obtain a token for the app user. 

The context is the multi user social media hub? so basically I would want to have several users stored in the db one way or the other

**Comment by cavallium** [comment]
Are you talking about sessions?

**Comment by Schuyweiz** [comment]
> Are you talking about sessions?

Im not sure if some sort of a token is stored in a tg user session. Anything that would allow to identify the user and allow to make new requests.
As far as I understand, all the clients are stored in a concurrent map in a singleton of InnerClientManager or something like that (will have a look, once at the pc). And I cant really add or extract clients from there. So it feels like unless i store it in a hasmap of some sort, I cannot restore the client object after application shutdown.

**Comment by Schuyweiz** [comment]
Maybe I should clarify further what I am looking for.
The documentation states "As a result of authorization, the client key, auth_key_id, becomes associated with the user, and each subsequent API call with this key will be executed with that user's identity."

So there is an id, that will be associated with the user in question. However, I do not seem to find how to extract it and store. Nor do I see any way of using the code to recreate a client, associated with a particular user.

So my question is if that's even possible with tdlight or maybe I could make it work with my own implementation of say TelegramInterface

**Comment by cavallium** [comment]
With tdlight you should only worry about the directory in which you created the session.
If you already logged in, the next time you open the session in your specified directory, no login will be requested to you.
So you should just store the locations of each session in a configuration file or something similar.

**Comment by Schuyweiz** [comment]
> With tdlight you should only worry about the directory in which you created the session.

But how do I initialize the client then? I can only specify in parameters the path go directory where a local sql db is initalized, but when I try to pass the same path, it throws an exception, that I cannot initialize a client with the same path to a local directory

**Comment by cavallium** [comment]
Please provide some code, I can't understand what you are trying to say
